pkg/ipmi/oem.go

```diff
@@ -0,0 +1,298 @@
+}
+
+func (i *IPMI) IpmiOemSetProcessorInfo(si *smbios.Info) error {
+	info := &ProcessorInfo{}
+
+	t1, err := si.GetSystemInfo()
+	if err != nil {
+		return err
+	}
+
+	t4, err := si.GetProcessorInfo()
+	if err != nil {
+		return err
+	}
+
+	boardManufacturerID, ok := BoardManufacturer[t1.Manufacturer]
+	if ok {
+		info.ManufacturerID = boardManufacturerID
+	}
+
+	for index := 0; index < len(t4); index++ {
+		info.Index = uint8(index)
+		copy(info.ProductName[:], t4[index].Version)
+		info.Index = uint8(index)
+		info.CoreNumber = uint8(t4[index].GetCoreCount())
+		info.ThreadNumberLSB = uint8(t4[index].GetThreadCount() & 0x00ff)
+		info.ThreadNumberMSB = uint8(t4[index].GetThreadCount() >> 8)
+		info.ProcessorFrequencyLSB = uint8(t4[index].MaxSpeed & 0x00ff)
+		info.ProcessorFrequencyMSB = uint8(t4[index].MaxSpeed >> 8)
+		info.Revision1 = 0
+		info.Revision2 = 0
+
+		for param := 1; param <= 2; param++ {
+			infodata, err := info.marshall(param)
+			if err != nil {
+				return err
+			}
+
+			err = i.SendOemIpmiCommand(_IPMI_FB_OEM_NET_FUNCTION, _FB_OEM_SET_PROC_INFO, infodata)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (i *IPMI) IpmiOemSetDimmInfo(si *smbios.Info) error {
+	info := &DimmInfo{}
+
+	t1, err := si.GetSystemInfo()
+	if err != nil {
+		return err
+	}
+
+	t17, err := si.GetMemoryDevices()
+	if err != nil {
+		return err
+	}
+
+	boardManufacturerID, ok := BoardManufacturer[t1.Manufacturer]
+	if ok {
+		info.ManufacturerID = boardManufacturerID
+	}
+
+	for index := 0; index < len(t17); index++ {
+		info.Index = uint8(index)
+		info.Index = uint8(index)
+		info.Index = uint8(index)
+		info.Index = uint8(index)
+		info.Index = uint8(index)
+		info.Index = uint8(index)
+
+		if t17[index].AssetTag == "NO DIMM" {
+			info.DIMMPresent = 0xFF // 0xFF - Not Present
+		} else {
+			info.DIMMPresent = 0x01 // 0x01 - Present
+		}
+
+		data := strings.Split(strings.TrimPrefix(t17[index].BankLocator, "_"), "_")
+		dimm, _ := strconv.ParseUint(strings.TrimPrefix(data[2], "Dimm"), 16, 8)
+		channel, _ := strconv.ParseUint(strings.TrimPrefix(data[1], "Channel"), 16, 8)
+		node, _ := strconv.ParseUint(strings.TrimPrefix(data[0], "Node"), 16, 8)
+		info.DIMMNumber = uint8(dimm)
+		info.ChannelNumber = uint8(channel)
+		info.NodeNumber = uint8(node)
+
+		infodata, err := info.marshall(1)
+		if err != nil {
+			return err
+		}
+
+		err = i.SendOemIpmiCommand(_IPMI_FB_OEM_NET_FUNCTION, _FB_OEM_SET_DIMM_INFO, infodata)
+		if err != nil {
+			return err
+		}
+
+		if info.DIMMPresent == 0x01 {
+			// DIMM type: bit[7:6] for DDR3 00-Normal Voltage(1.5V), 01-Ultra Low Voltage(1.25V), 10-Low Voltage(1.35V), 11-Reserved
+			//                     for DDR4 00~10-Reserved, 11-Normal Voltage(1.2V)
+			//            bit[5:0] 0x00=SDRAM, 0x01=DDR1 RAM, 0x02-Rambus, 0x03-DDR2 RAM, 0x04-FBDIMM, 0x05-DDR3 RAM, 0x06-DDR4 RAM
+			if t17[index].Type == smbios.MemoryDeviceTypeDDR3 {
+				switch t17[index].ConfiguredVoltage {
```

Comment:
would it make sense to have separate functions for this conversion?

Reply:
Do you mean have a function like "DetectDimmType"?

Reply:
Done.